source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,411 @@
+  // LbSubsetEntry for the criteria, which may or may not have a
+  // subset attached to it.
+  for (uint32_t i = 0; i < match_criteria.size(); i++) {
+    const Router::MetadataMatchCriterion& match_criterion = *match_criteria[i];
+    const auto& subset_it = subsets->find(match_criterion.name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match_criterion.value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (i + 1 == match_criteria.size()) {
+      // We've reached the end of the criteria, and they all
+      // matched.
+      return entry;
+    }
+
+    subsets = &entry->children_;
+  }
+
+  return nullptr;
+}
+
+// Given the addition and/or removal of hosts, update all subsets,
+// creating new subsets as necessary.
+void SubsetLoadBalancer::update(const std::vector<HostSharedPtr>& hosts_added,
+                                const std::vector<HostSharedPtr>& hosts_removed) {
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+    HostPredicate predicate = [](const Host&) -> bool { return true; };
+    if (fallback_subset_ == nullptr) {
+      // First update: create the default host subset with all hosts
+      fallback_subset_.reset(new LbSubsetEntry());
+      fallback_subset_->initLoadBalancer(*this, predicate);
+    } else {
+      // Subsequent updates: add/remove hosts
+      fallback_subset_->host_subset_->update(hosts_added, hosts_removed, predicate);
+    }
+  } else if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET) {
+    HostPredicate predicate =
+        std::bind(&SubsetLoadBalancer::hostMatchesDefaultSubset, this, std::placeholders::_1);
+    if (fallback_subset_ == nullptr) {
+      // First update: create the default host subset by filtering for hosts that match the criteria
+      fallback_subset_.reset(new LbSubsetEntry());
+      fallback_subset_->initLoadBalancer(*this, predicate);
+    } else {
+      if (!hosts_added.empty() || !hosts_removed.empty()) {
+        fallback_subset_->host_subset_->update(hosts_added, hosts_removed, predicate);
+      }
+    }
+  }
+
+  std::unordered_set<LbSubsetEntryPtr> subsets_modified;
+
+  for (const auto& host : hosts_added) {
+    for (const auto& keys : subset_keys_) {
+      // For each added host, for each subset key, attempt to extract the metadata
+      // corresponding to the key from the host.
+      SubsetMetadata kvs = extractSubsetMetadata(keys, *host);
+      if (!kvs.empty()) {
+        // The host has metadata for each key, find or create its subset.
+        LbSubsetEntryPtr entry = findOrCreateSubset(subsets_, kvs, 0);
+        if (entry != nullptr) {
```

Comment:
Give we say above that the plan is to "fine or create its subset", how can this now be `nullptr`? I'd recommend a one-liner explaining how it could be this way, or an `ASSERT` stating it can't.